src/sinks/statsd.rs

```diff
@@ -201,6 +202,7 @@ fn encode_event(event: Event, default_namespace: Option<&str>) -> Option<Vec<u8>
                 StatisticKind::Histogram => "h",
                 StatisticKind::Summary => "d",
             };
+            let samples = compress_distribution(samples.clone());
```

Comment:
While `into_parts` would assign to `data` by value, `event` would be moved, and then we have no other choice but to call `push_event(..., event.clone.as_metric(), ...)` (or something like that) later in the `match data.value` matching arms. I don't think we can avoid a `clone()` here, but please let me know if I am wrong. `samples.clone()` is probably the most lightweight clone() among all possibilities.

Reply:
Good point, I had glossed over all the other uses of `metric` below. I think it is possible to achieve this without a `clone`, but it will require some reworking of `compress_distribution` to take `samples` by (mutable) reference and do the compression in place. That actually makes it a candidate to be a method on the metric types instead of scoped to the metric buffer.

If you are interested in doing that work, I'd be happy to accept it as part of this PR. Otherwise, I'll add a note for us to work on it later and this can go in as-is.

Reply:
@bruceg I would love to help implement what you suggested, but `metric` is already borrowed in `match metric.value()`, so if we want to call `metric.compress_distribution` (which takes &mut self), `metric.value()` needs to return a copy instead of a reference. I don't think it's a good idea to modify the returned type of `value()`. WDYT?

Reply:
Right, so we would need a `Metric::value_mut(&mut self) -> &mut MetricValue` to get the mutable reference to use in the `match`. Then the relevant arm would get a mutable reference to `samples`. If the borrow checker doesn't like that, we might have to separate out the compression into a phase before the encoding `match`, ie:
```rust
if let MetricValue::Distribution { mut ref samples, ... } = metric.value_mut() {
    compress_distribution(samples);
}

match metric.value() {
    …
}
```
(not sure if the `mut ref` is necessary there)

Reply:
To make use of `Metric::value_mut(&mut self) `, many changes has to be propagated, but such change may not make sense e.g. change `fn encode_event(event: Event, default_namespace: Option<&str>)` to `fn encode_event(event: mut Event, default_namespace: Option<&str>)`. I would propose to "eliminate the `.clone`" in a separate PR, so please merge if you are ok with this.

Reply:
I am fine with that. I agree it is a bigger change than warranted for this PR.